### ISSUE TITLE
Calling `spookfox-org-save-all-tabs` doesn't save tabs from every window

### ISSUE THREAD

**Issue by seanhagen** [bitspook/spookfox#35]
I can see that in `spookfox-org-tabs--request-all-tabs`, you only get the first client from `spookfox--connected-clients` using `cl-first`. 

I tried modifying `spookfox-org-tabs--request-all-tabs` like so:

```elisp
(defun spookfox-org-tabs--request-tabs-from-client (client)
  "Gets all tabs from CLIENT."
  (when client
    (plist-get
     (spookfox--poll-response
      (spookfox-org-tabs--request client "GET_ALL_TABS"))
     :payload)))

(defun spookfox-org-tabs--request-all-tabs ()
  "Get all tabs from each connected client."
  (let ((client-list spookfox--connected-clients)
        (tabs ()))
    (while client-list
      (setq tabs (cons (spookfox-org-tabs--request-tabs-from-client (cdr client-list)) tabs))
      (setq client-list (car client-list)))
    tabs))
```

I was thinking if I just loop through the connected clients, get the tabs from that client and append them to a list -- then I could return the list of tabs from all clients and it'd be golden!

No luck though, I get errors:
```
spookfox-org-tabs--request-tabs-from-client: Wrong type argument: listp, #s(websocket open nil spookfox--handle-new-client spookfox--handle-msg (closure ((user-method . spookfox--handle-disconnect-client)) (ws) (setq websocket-server-websockets (remove ws websocket-server-websockets)) (funcall user-method ws)) spookfox--handle-server-error nil nil t ...)
```

I'm not super great with elisp, and I'm also hacking on this while trying to keep an eye on work stuff 😅 so I'm probably missing something pretty basic in my code.

However, if instead `spookfox-org-tabs--request-all-tabs` is defined like so:

```elisp
(defun spookfox-org-tabs--request-all-tabs ()
  "Get all tabs currently present in browser."
  (let ((client (cl-first spookfox--connected-clients)))
    (spookfox-org-tabs--request-tabs-from-client client)))
```

Calling `spookfox-org-tabs-save-all-tabs` works fine; it just still only shows the tabs from a single window. So the issue with my code has to do with the version that uses `while` and how it constructs the list of tabs, I think.

Alternatively if you don't want the code to loop through each client, then maybe if there's more than one client could the code let the user select which client to pull tabs from?

I always have multiple Firefox windows open, so being limited to only the first window for saving tabs is not ideal.

**Comment by bitspook** [comment]
Hello @seanhagen! 

tldr:
You can access tabs from all windows of firefox using `spookfox--request-all-tabs`. This function comes from `spookfox-tabs` app. I just pushed this change in [this commit](https://github.com/bitspook/spookfox/commit/1b52fcd7e827d2048f192c2862e4bc00ae7255d8). Now `spookfox-org-tabs--request-all-tabs` should give you tabs from all browser windows.

Detail:

I'd change your code slightly for easier iteration:

```lisp
(defun hypothetically-get-tabs-of-client ()
  (let ((client-list spookfox--connected-clients)
      (tabs nil))
  (dolist (client client-list)
    (setf tabs (append tabs (hypothetically-get-tabs-of-client client))))
  tabs))
```

`dolist` is a bit easier to handle imo. 

This should fix your immediate problem, but will not give you list of all tabs from all browser windows. Unfortunately, multiple windows of same browser don't register as new clients in `spookfox--connected-clients`. Only entirely new instances of the browser would initiate new connection, and hence become new `client`s. 


> if you don't want the code to loop through each client

This is not the case. Supporting more clients is a good idea I think, it is however not the one which is implemented. 

> if there's more than one client could the code let the user select which client to pull tabs from

That would be ideal. 

Iterating over all clients or allowing user to select a client will not allow us to achieve what you've requested, but it will enable following:
1. Supporting multiple browsers. We will be able to select tabs from both FF and Chromium in same place.
2. Supporting multiple instances of a browser. Opening a new window from running browser isn't considered a new instance. A new instance would be created e.g if we run `firefox` from 2 different terminals. 

This means to get all tabs from all windows of same browser, we need do one of these:
1. Get list of all windows from browser, and request tabs for every window
2. Ask browser to give us *all* tabs and not just tabs of a single window. Which is what [spookfox-tabs](https://github.com/bitspook/spookfox/blob/1b52fcd7e827d2048f192c2862e4bc00ae7255d8/spookfox-addon/src/apps/Tabs.ts#L76) app do.

---
PS I'd advise not to invest too much in spookfox-org-tabs app. I will quite likely [pivot it](https://github.com/bitspook/spookfox/issues/30) soon.

**Comment by seanhagen** [comment]
Thanks for the update! What you have in that pivot ticket is pretty much what I want to do anyways, so I'm looking forward to those changes 😄
